library/core/src/pin.rs

```diff
@@ -110,79 +221,221 @@
+//!             if self.start.is_null() {
+//!                 self.start = node;
+//!             } else {
+//!                 (*self.end).succ = node;
+//!                 node.pred = self.end;
+//!             }
+//!             self.end = node;
+//!         }
+//!     }
+//!
+//!     /// Allocates a node on the heap and appends it to the end of the list.
+//!     fn append_boxed(&mut self) -> Pin<Box<Node>> {
+//!         let mut node = Box::pin(Node {
+//!             pred: ptr::null_mut(),
+//!             succ: ptr::null_mut(),
+//!             data: Data::new(),
+//!             _pin: PhantomPinned,
+//!         });
+//!         self.append(Pin::as_mut(&mut node));
+//!         node
+//!     }
+//! }
+//!
+//! # struct Data;
+//! # impl Data { fn new() -> Self { Data } }
+//! struct Node {
+//!     pred: *mut Node,
+//!     succ: *mut Node,
+//!     data: Data,
+//!     /// `Node: Unpin` because `List` expects `Pin`ned pointers to them to remain in place.
+//!     _pin: PhantomPinned,
+//! }
+//!
+//! impl Drop for Node {
+//!     /// Remove pointers to `self`, allowing reuse of this memory without clients seeing garbage.
+//!     fn drop(&mut self) {
+//!         if self.pred.is_null() || self.succ.is_null() {
+//!             return
+//!         }
+//!
+//!         unsafe {
+//!             (*self.pred).succ = self.succ;
+//!             (*self.succ).pred = self.pred;
+//!         }
+//!     }
+//! }
+//! ```
+//!
+//! For this to work, a [`drop`-related guarantee][drop-guarantee] is required. If an element could
+//! be deallocated or otherwise invalidated without calling [`drop`], the pointers into it from its
+//! neighboring elements would become invalid, which would break the data structure.
+//!
+//! [`List`] itself is *not* address-sensitive.
+//!
+//! # Subtle details
+//! [subtle-details]: #subtle-details
+//!
+//! [`List::append`] above relies on both of [`Pin<P>`]'s guarantees:
+//!
+//! <!-- NOTE from anp: "until further notice" is awkward phrasing but i assume you've thought a bit
+//!      about it. consider "until destruction allows fixing up pointers" or similar? -->
 //!
-//! In an intrusive doubly-linked list, the collection does not actually allocate
-//! the memory for the elements itself. Allocation is controlled by the clients,
-//! and elements can live on a stack frame that lives shorter than the collection does.
+//! 1.  *Address Stability.* If [`unsafe`] code witnesses any [`p: Pin<P>`][Pin] at any time then
+//!     it may assume that `p.as_ref().get_ref() as *const _` will remain valid, pointing to the
+//!     same object until further notice.
+//! 2.  *Notice of Destruction.* If `x: T` was ever reachable through any [`Pin<P>`] type, its
+//!     destructor must be run (until it either returns or panics) before `x`'s storage can be
+//!     overwritten. The "until further notice" in (1) includes this mandatory destruction. This is
+//!     often called the "[`Drop`] guarantee".
```

Comment:
Might be worth mentioning that you don't have to drop it if you leak the memory.

Reply:
This is noted in the larger section on the drop guarantee. I think it's a bit of a footnote more than anything else.